exercises/concept/poetry-club-door-policy/door-policy.spec.js

```diff
@@ -7,239 +9,117 @@ import {
+    test('should take the first character of each line', () => {
+      const key = recite(SHIRE_HORSE, frontDoorPatron);
+      const expectedFirstLetters = ['S', 'H', 'I', 'R', 'E'];
+      expect(key).toEqual(expectedFirstLetters);
     });
 
-    test('it generates the correct password', () => {
-      ShireGuard.assert();
+    test('should generate the correct password', () => {
+      expect(frontDoorPassword('SHIRE')).toBe('Shire');
     });
 
-    const FRONT_DOOR_CASES = {
-      SUMMER,
-      SOPHIA,
-      CODE_WORK,
-    };
-
-    Object.keys(FRONT_DOOR_CASES).forEach((name) => {
-      const poem = FRONT_DOOR_CASES[name];
-      const guard = new FrontDoorGuard(poem);
-
-      test(`frontDoorPassword(${name})`, () => {
-        guard.assert();
-      });
-    });
+    test.each([
+      { poem: SUMMER, firstLetters: 'SUMMER', password: 'Summer' },
+      { poem: SOPHIA, firstLetters: 'SOPHIA', password: 'Sophia' },
+      { poem: CODE_WORK, firstLetters: 'CODE', password: 'Code' },
+    ])(
+      'should be correct for $password',
+      ({ poem, firstLetters, password }) => {
+        expect(recite(poem, frontDoorPatron).join('')).toBe(firstLetters);
+        expect(frontDoorPassword(firstLetters)).toBe(password);
+      }
+    );
   });
 
-  describe('back door', () => {
-    const ShireGuard = new BackDoorGuard(SHIRE_HORSE);
-
-    test('it outputs a character per line', () => {
-      const key = ShireGuard.recite();
+  describe('back door response', () => {
+    test('should output a character per line', () => {
+      const key = recite(SHIRE_HORSE, backDoorPatron);
       expect(key.length).toBe(SHIRE_HORSE.length);
     });
 
-    test('it outputs takes the first characters', () => {
-      const key = ShireGuard.recite();
-      expect(key.toUpperCase()).toBe(SHIRE_HORSE.telestich.toUpperCase());
+    test('should take the last letter character of each line', () => {
+      const key = recite(SHIRE_HORSE, backDoorPatron);
+      expect(key).toEqual(['h', 'o', 'r', 's', 'e']);
     });
 
-    test('it generates the correct password', () => {
-      ShireGuard.assert();
+    test('should generate the correct password', () => {
+      expect(backDoorPassword('horse')).toBe('Horse, please');
     });
 
-    const BACK_DOOR_CASES = {
-      CODE_WORK,
-      SHIRE_HORSE_WITH_SPACES,
-    };
-
-    Object.keys(BACK_DOOR_CASES).forEach((name) => {
-      const poem = BACK_DOOR_CASES[name];
-      const guard = new BackDoorGuard(poem);
-
-      test(`backDoorGuard(${name})`, () => {
-        guard.assert();
-      });
+    test.each([
+      { poem: CODE_WORK, lastLetters: 'work', password: 'Work, please' },
+      {
+        poem: SHIRE_HORSE_WITH_SPACES,
+        lastLetters: 'horse',
+        password: 'Horse, please',
+      },
+    ])('should be correct for $password', ({ poem, lastLetters, password }) => {
+      expect(recite(poem, backDoorPatron).join('')).toBe(lastLetters);
```

Comment:
Because of how this test works with the join, all tests will pass even if the student does not use `trim` in `backDoorResponse`. The tests should catch the missing `trim`, e.g. by also running `SHIRE_HORSE_WITH_SPACES` through the "output a character per line" test.

Reply:
Restructured the tests to lose the `join()` and added a test using `SHIRE_HORSE_WITH_SPACES`.